### ISSUE TITLE
PublishTrimmed should imply self-contained

### ISSUE THREAD

**Issue by eerhardt** [dotnet/sdk#30104]
Doing the following steps:

* `dotnet new console`
* `dotnet publish -p:PublishTrimmed=true`

Results in the following error:

```
C:\Users\eerhardt\.nuget\packages\microsoft.net.illink.tasks\8.0.100-1.22619.1\build\Microsoft.NET.ILLink.targets(196,5): error NETSDK1102: Optimizing assemblies for size is not supported for
the selected publish configuration. Please ensure that you are publishing a self-contained app. [C:\DotNetTest\Net8Console\Net8Console.csproj]
```

However, doing

* `dotnet publish -p:PublishAot=true`
* -or-
* `dotnet publish -p:PublishSingleFile=true`

Just works without more command line args.

We should make `PublishTrimmed=true` imply `SelfContained=true`, just like `PublishAot=true` and `PublishSingleFile=true` does. That way users don't need to pass extra args on the command line, just to make this work.

If in the future we want to support a mode where `PublishTrimmed=true`, but `SelfContained=false`, those customers can explicitly `--self-contained false`.  This is consistent with the current `PublishSingleFile=true` behavior. By default it is self-contained. And you can opt into Framework Dependent Deployment with `dotnet publish -p:PublishSingleFile=true --self-contained false`.

cc @vitek-karas @sbomer @agocke

**Comment by agocke** [comment]
Surprised that `dotnet publish -p:PublishSingleFile=true` is self-contained by default. But I don't dislike the behavior. And agree with the proposal about PublishTrimmed.

The only thing I don't like is that these properties affect `dotnet build` as well. In fact, I don't like self-contained `dotnet build` at all. I think that should be a publish-only gesture.

**Comment by marcpopMSFT** [comment]
@sbomer see https://github.com/dotnet/sdk/pull/30152. Feel free to update this in the future if there are changes needed.

**Comment by richlander** [comment]
Good topic! We're trying to provide coherent behaviors for some things that are similar and others less so. It's sometimes easier when we group the options such that the answer pops out. Let's see if that happens here. 

I'm going to use MSBuild syntax, where all properties are intended as `true` or otherwise have coherent value.

Always framework-dependent:

- [nothing]
- `RuntimeIdentifier`

Supports both:

- `PublishReadyToRun`
- `PublishSingleFile`

Could support both, but only supports self-contained:

- `PublishTrimmed`

Always self-contained:

- `PublishNativeAot`
- `PublishSelfContained`
- `SelfContained`

There are two ways we could go with this:

- Correctness
- Convenience

There are four groups. The pragmatic choice is make the first two groups fx-dependent by default and the second two are self-contained.

Another observation is that `PublishTrimmed` is likely not used that often on its own. Instead, we could do the following:

- Require users to specifying self-contained if they use `PublishTrimmed`
- Default `PublishSelfContained` to infer `PublishTrimmed` and require users to disable trimming if they don't want it. We could also infer `PublishReadyToRun` in the same way.

That approach would have multiple benefits:

- Most of the options would be neutral to the output type (which is good).
- We'd have an option for producing self-contained apps that is good. Most people would agree that the current default for self-contained apps is not polished/optimal.
- Single file would remain a separate concern. The thinking to date is that the various forms of single file should be opt-in since there are more compat considerations with single file variants than with self-contained.

This approach seems to balance CLI consistency with convenience while also most often providing users with a set of artifacts that we would say are "good".

**Comment by agocke** [comment]
@richlander Based on my comment above, I want to retire `SelfContained` as it is today. Specifically, I think if you want a self-contained output you should have to use `dotnet publish`.

**Comment by agocke** [comment]
A couple more things, actually.

> Default PublishSelfContained to infer PublishTrimmed and require users to disable trimming if they don't want it. We could also infer PublishReadyToRun in the same way.

I'm not sure I understand what you're suggesting here. You mean that if `PublishSelfContained` is true, the default for `PublishTrimmed` is also true? Or the other way around?

> Single file would remain a separate concern. The thinking to date is that the various forms of single file should be opt-in since there are more compat considerations with single file variants than with self-contained.

The compat between self-contained single file and plain self-contained is getting pretty good at this point. It's pretty much just Assembly.Location now.

**Comment by richlander** [comment]
That's a good point @sbomer. Let's go with that.

I'm thinking about this more. I was trying to get at something with that categorization in my previous post, but I didn't quite know what. I do now.

Let's try again.

App characteristics (don't auto-switch from FDD -> SCD):

- `PublishReadyToRun`
- `PublishSingleFile`
- `PublishTrimmed` 
- `RuntimeIdentifier`
 
App kind (these auto switch to SCD):
- `PublishNativeAot`
- `PublishSelfContained`
- `SelfContained`

This feels like a more useful categorization, but I still don't find it particularly satisfying. For example, `PublishSingleFile` feels like it should be an app kind and `PublishTrimmed` only supports self-contained apps currently.

In the ideal case, users set the minimum values in their project file, like setting `PublishSelfContained=true`. If they want to try out trimming or single file from the CLI, they will always get the desired result.

I think this conversation comes down to specializing a default project file (like what templates produce) via the CLI. I think a solution like this would help: https://github.com/dotnet/sdk/issues/26249#issuecomment-1406940259. It's basically the same thing as specifying the same values in a project file.

@agocke -- `SelfContained` ... I'm a little torn on that one. We need to decide once and for all that we're going to stay with `publish` forever. If we agree to that, then I'm happy to go along with your suggestion. Related: https://github.com/dotnet/sdk/issues/26247.

@agocke -- I meant that `PublishSelfContained` could infer `PublishTrimmed`. I had proposed single file as a default in the past. @jkotas considered that too much of a compatibility risk for opt-out.

Upon further thinking, I think we should go with the categorization I have above and then rely on good project file and CLI gestures to make it easy to adopt the configuration you want. We just added `PublishSelfContained` and the bad PM for that feature hasn't published a blog post on that yet. I think we don't have enough customer feedback to suggest we need something more dramatic. I'd rather see us make the CLI easier to use to specify defaults for an environment.

**Comment by richlander** [comment]
@baronfel @dsplaisted

**Comment by nagilson** [comment]
Thanks everyone for engaging here + internally 😄 

@richlander Should we still take this change: https://github.com/dotnet/sdk/pull/30038 if it does not have `PublishAot` giving `SelfContained` by default? Today is the last day for it to get merged (technically Monday at 3 PM...). It sounds like we are OK with this change breaking `PublishTrimmed`, `PublishSingleFile`, and `PublishReadyToRun` such that old usages of these will now have to add `SelfContained`.

If we have a finalized decision on this, we should close/edit this issue since we've seem to have decided against it, and make one specifically for having `PublishAot` -> `SelfContained`, as well as one for `PublishSelfContained`.

**Comment by richlander** [comment]
Right. I'm suggesting that `PublishAot` and `PublishSelfContained` should infer `SelfContained`.

I think all the other cases should not infer a `SelfContained` value. That breaks our model, IMO, and is impossible to explain w/o giving folks a decoder ring.

**Comment by richlander** [comment]
Up-leveling a bit. We should be dissatisfied with experiences that are inconvenient. However, addressing those at the lowest levels (like with these properties) is just one option. We should ensure we have all the options on the table for how to address UX. I'm proposing we handle this UX concerns (like `PublishTrimmed` being annoying at a higher level. It's really only annoying for the CLI scenario.

**Comment by richlander** [comment]
I don't think `PublishAot` and `PublishTrimmed` are very similar. We could enable `PublishTrimmed` for FDD if we wanted to. We just haven't. It also means we're one motivating customer request from doing so.

Another middling option is to expose another set of CLI flags, like `--trim-scd` or similar.

**Comment by jkotas** [comment]
> We could enable PublishTrimmed for FDD if we wanted to. We just haven't. 

The same can be said about PublishAot. (FDD was the default deployment option for .NET Native for UWP.)

**Comment by richlander** [comment]
You are right! I forgot about that. However (and correct me if I'm wrong), it seems more likely that we'll enable trimming for FDD than Native Aot for FDD. Or, do we just need to wait on a similar "one motiving scenario"?

**Comment by jkotas** [comment]
I agree.

**Comment by richlander** [comment]
That's a bit ambiguous. There were two things you could agree with. I assume you were agreeing that trimming for FDD was more likely.

**Comment by jkotas** [comment]
> I assume you were agreeing that trimming for FDD was more likely.

Yep.

**Comment by richlander** [comment]
Here's a concrete proposal, since we still have time on our side.

- Adopt a more neutral approach with these properties (only `PublishNativeAOT` and `*SelfContained` imply SDC).
- Agree that project files have sufficiently descriptive means to communicate the desired outcome efficiently and effectively.
- Switch focus to improving the CLI to make it similarly efficient and effective.

I make no secret that I think this approach would be a good solution for this specific topic area and also have broader utility for even more common scenarios.

**Comment by vitek-karas** [comment]
As stated above there's a conflict between consistency and "ease of use". The above discussion makes it clear that we're proposing sacrifying "easy of use on CLI" and want consistent CLI options. (Personally I'm split on this, but if most people think that's the right approach, let's do it).
This means that for example adding trimming in CLI will be somewhat cumbersome (need to also specify SCD).

With that, I don't see why we would special case `PublishNativeAOT` as the only which implies SCD. On the scale of "basic" to "advanced" uses, AOT is probably the most advanced right now, so we could argue that making it easy to use from CLI is low priority.

Another way to look at it - if we think we can solve the UX for trimming outside of CLI, I don't see why we could not solve it the same way for AOT as well. And then keep the CLI 100% consistent.

As already mentioned, we could introduce CLI-only aliases which solve the UX:
We're already introducing `dotnet new api -aot` and thus it would make sense to have `dotnet publish -aot` which will "do the right thing" (and that could mean different things for different TFMs if we really needed to). Basically, decouple the MSBuild properties (very explicit and orthogonal/independent) from use-of-use CLI options (targeted at the most frequent scenarios).

**Comment by nagilson** [comment]
`PublishSelfContained` by default is an interesting idea. Another idea per @marcpopMSFT is that we could infer `SelfContained` by default when using these properties, BUT have a warning like the old SelfContained warning, 


```
WARNING: NETSDK11??: One of '--self-contained' or '--no-self-contained' options are recommended when `Publish*` is used.
```

Thoughts @richlander ?

**Comment by eerhardt** [comment]
>> I don't think making a breaking change for PublishSingleFile to default to framework-dependent is correct

> I agree that this is one of the biggest downsides to the breaking change (RID no longer implying self-contained). Same applies to vanilla "dotnet publish -r <RID>".

IMO - This isn't part of the RID breaking change. I didn't specify a RID at all: `dotnet publish -p:PublishSingleFile=true`. I don't get a warning on .NET 7, like I do for `-r <RID>`.

**Comment by richlander** [comment]
Here's the product behavior to date. It's useful to record that so that we're all on the same page.

.NET 6:

```bash
root@d2e8b3f38232:/app# dotnet --version
6.0.405
root@d2e8b3f38232:/app# dotnet new console -o app
root@d2e8b3f38232:/app# dotnet publish
root@d2e8b3f38232:/app# find . | grep coreclr
root@d2e8b3f38232:/app# dotnet publish -p:SelfContained=true
MSBuild version 17.3.2+561848881 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
/usr/share/dotnet/sdk/6.0.405/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(154,5): error NETSDK1031: It is not supported to build or publish a self-contained application without specifying a RuntimeIdentifier. You must either specify a RuntimeIdentifier or set SelfContained to false. [/app/app.csproj]
```

.NET 7:

```bash
root@d94e82f2a8a2:/app# dotnet --version
7.0.102
root@d94e82f2a8a2:/app# dotnet new console
root@d94e82f2a8a2:/app# dotnet publish
root@d94e82f2a8a2:/app# find . | grep coreclr
root@d94e82f2a8a2:/app# dotnet publish -p:SelfContained=true
MSBuild version 17.4.1+9a89d02ff for .NET
  Determining projects to restore...
  Restored /app/app.csproj (in 4.97 sec).
  app -> /app/bin/Debug/net7.0/linux-x64/app.dll
  app -> /app/bin/Debug/net7.0/linux-x64/publish/
root@d94e82f2a8a2:/app# find . | grep coreclr
./bin/Debug/net7.0/linux-x64/libcoreclr.so
root@d94e82f2a8a2:/app# dotnet publish -p:PublishTrimmed=true
MSBuild version 17.4.1+9a89d02ff for .NET
  Determining projects to restore...
  Restored /app/app.csproj (in 64 ms).
  app -> /app/bin/Debug/net7.0/app.dll
/usr/share/dotnet/sdk/7.0.102/Sdks/Microsoft.NET.ILLink.Tasks/build/Microsoft.NET.ILLink.targets(196,5): error NETSDK1102: Optimizing assemblies for size is not supported for the selected publish configuration. Please ensure that you are publishing a self-contained app. [/app/app.csproj]
root@d94e82f2a8a2:/app# dotnet publish -p:PublishSingleFile=true
MSBuild version 17.4.1+9a89d02ff for .NET
  Determining projects to restore...
  Restored /app/app.csproj (in 104 ms).
  app -> /app/bin/Debug/net7.0/linux-x64/app.dll
  app -> /app/bin/Debug/net7.0/linux-x64/publish/
root@d94e82f2a8a2:/app# dotnet publish -p:PublishNativeAot=true
MSBuild version 17.4.1+9a89d02ff for .NET
  Determining projects to restore...
  Restored /app/app.csproj (in 81 ms).
  app -> /app/bin/Debug/net7.0/app.dll
  app -> /app/bin/Debug/net7.0/publish/
```

I actually thought we implemented `PublishSelfContained` last version, but I guess not. As I recall, this isn't the first time I've been confused by that.

The quick summary is this:

-  `SelfContained=true` implies a RID and previously did not.
- `PublishTrimmed` doesn't imply a deployment type
- `PublishAot` and `PublishSingleFile` imply SCD.

**Comment by richlander** [comment]
> However, I would find it really unexpected if "dotnet publish" produced one kind (SCD/FDD) of deployment, but "dotnet publish -p:PublishSingleFile=true" produced a different kind.

It is even more true for `dotnet build` to `dotnet publish`, in general. We cannot change `dotnet publish` to be SCD. That's a very large breaking change and would fundamentally change the product. For example, there are tons of people that use our container images. They need to produce FDD apps for that to make sense.

I guess after all of this conversation, I conclude that we should leave the behavior as-is in .NET 7. I think we all agree that there is an affinity to SCD with Single File and AOT. That's why they have the current behavior. I'm no longer proposing we change that.

I'm still not convinced on `PublishTrimmed`. We are going through a painful multi-release effort of changing what `dotnet publish -r some-rid` means, making it FDD by default.  I don't want to regret this moment when we later decide to support FDD trimming. I suspect it will happen at some point.

Let's assume that we were enabling FDD trimming this release. What would we do?

I think we'd enable it for both FDD and SCD `PublishTrimmed`, and (like now) it wouldn't imply either.

We'd then need to decide if we wanted a simpler way to enable `PublishTrimmed` + SCD in one gesture. I'd be happy with `PublishTrimmedScd`. Is that good enough to satisfy the usability concerns?

Tell me if I'm wrong, but I don't think we need to over-prioritize the loose-files scenario. I'm guessing that most folks prefer one of our single file scenarios. Also, has there been many requests to improve this? If not, then that suggests to me that people are using the single file options.

**Comment by richlander** [comment]
I just did some more research. I forgot about `PublishReadyToRun`. It implies a RID and SCD. I updated the comment above to include that.

I'm not sure I agree with my previous thinking (in .NET 7) that led to this point, but I'm now coming to a different conclusion given the experiences all pushing one way and a desire for consistent behaviors and not breaking compat.

Back to what we'd do if we were enabling `PublishTrimmed` now. I'd suggest then that we consider adding a `PublishTrimmedFdd` variant and same to the other `Publish*` properties that currently imply SCD but also can support FDD.

Outside of that idea, people could update their project files like this and that would work.

```xml
<PropertyGroup>
  <PublishTrimmed>true</PublishTrimmed>
  <PublishSelfContained>false</PublishSelfContained>
</PropertyGroup>

That leads me to the conclusion that all the `Publish*` properties should imply SCD and a RID.

**Comment by richlander** [comment]
Unless we change it. We need to decide on the intended behavior and then we have deliver it. I have made a proposal on that.

**Comment by DamianEdwards** [comment]
Random thought, might be nice to enable something like this (it would just set the specific properties during the relevant target, with values for the specific properties winning if also specified):

```xml
<PropertyGroup>
  <PublishKind>Trimmed;SelfContained;SingleFile;Release</PublishKind>
</PropertyGroup>
```

The valid values are any of the '*' values from the 'Publish*' properties. If an invalid combination is specified it would error.

**Comment by agocke** [comment]
@DamianEdwards Meh. Sorry, seems like clutter to me, and now I have to remember two ways of configuring the flags. 😄

**Comment by DamianEdwards** [comment]
@agocke well that's just, like, your opinion man 😉 

But seriously, eye of the beholder and all that. But looking at this sea of publish properties made me desire something more concise, especially when we start talking about combination properties like `PublishTimmedScd` *ewww*

**Comment by agocke** [comment]
>  combination properties like PublishTimmedScd ewww

Yeah, I'm not convinced we should do those either. At the end of the day I think the verbosity of setting each individual flag isn't such a big deal, and much of the verbosity actually comes from XML. I'd rather not create complex parsing just to work around XML verbosity. Whether or not we like it, it's the language we've chosen.

**Comment by DamianEdwards** [comment]
>  I'd rather not create complex parsing just to work around XML verbosity.

I understood this was an established pattern in our build system already (semi-colon separated values) so I didn't consider it as introducing anything new or complex.

**Comment by agocke** [comment]
Items are generally preferred, because then you don't have the (ridiculously common, always infuriating) problem of `<NoWarn>CA12356</NoWarn>` accidentally overriding all the previous `<NoWarn>` clauses.

But the difficulty mostly comes from your suggestion of parsing the separated values and then using them to set the PublishX property. Not easy to implement in MSBuild target language and not possible in a build task.

**Comment by richlander** [comment]
> it would just set the specific properties during the relevant target

Interesting idea, but I don't think that's the key aspect we're struggling with. In fact, that feels a bit like an anti-pattern since we're trying to make the product provide best practice behavior by default. It just turns out that's hard in some cases..

> especially when we start talking about combination properties like PublishTimmedScd ewww

Agree.

> How are the publish options different? (Why should no-R2R be FDD, but R2R be SCD by default?)

You are right. The current system doesn't make much sense. When I was writing last night, I was getting "compat afraid". So, let's play the "compat be damned" game.

I'd switch all of these to FDD by default, give the teams the opportunity to provide a valuable FDD behavior, and then add a `--scd` flag to the CLI. Obviously `PublishSelfContained` would remain SCD.

Actually, I'd also leave `PublishNativeAot` as-is, implying SCD. I don't think we're ever going to provide an FDD behavior for it. As Jan raised, we've done that in the past, but I am extremely doubtful that we'll ever repeat that.

This release is definitely our last opportunity to make these changes. What do folks think of this proposal? Certainly, it would be the most coherent to the product overall.

**Comment by DamianEdwards** [comment]
> and then add a --scd flag to the CLI

We already have `--self-contained` though right? Or am I misunderstanding?

**Comment by richlander** [comment]
Yes. It's a pain to type. What I'm hearing from this conversation is that people on the team (who I totally recognize are not our customers) find the following to be arduous:

```bash
root@d94e82f2a8a2:/app# dotnet publish -p:PublishTrimmed=true
MSBuild version 17.4.1+9a89d02ff for .NET
  Determining projects to restore...
  Restored /app/app.csproj (in 64 ms).
  app -> /app/bin/Debug/net7.0/app.dll
/usr/share/dotnet/sdk/7.0.102/Sdks/Microsoft.NET.ILLink.Tasks/build/Microsoft.NET.ILLink.targets(196,5): error NETSDK1102: Optimizing assemblies for size is not supported for the selected publish configuration. Please ensure that you are publishing a self-contained app. [/app/app.csproj]
root@d94e82f2a8a2:/app# dotnet publish -p:PublishTrimmed=true --self-contained
```

The UX ask was to make `-p:PublishTrimmed=true` to be SCD. While also recognize my flip-flop-ness on the topic, I've been pushing back on that. I'm trying to make the product consistent and easier to use at the same time. Adding `--scd` was intended to help with that. Also, lots of CLIs offer short forms like that so it doesn't seem strange.

For example:

```bash
$ wget --help | grep verbose
  -v,  --verbose                   be verbose (this is the default)
  -nv, --no-verbose                turn off verboseness, without being quiet
```

**Comment by agocke** [comment]
> Actually, I'd also leave PublishNativeAot as-is, implying SCD

This feels like my argument about commonality-over-consistency. Even if we build PublishAot FDD, it will be much less common than scd. I think at can make the same argument for Publish Trimmed or single file

**Comment by richlander** [comment]
NativeAot is most that way and we get farther from that level of certainty as we go. They don't sit at the same spot on the spectrum. That's why I don't find this argument convincing.

Also, we're never going to build Native AOT FDD. It's a waste of time since no one wants that modality.

**Comment by agocke** [comment]
> Also, we're never going to build Native AOT FDD

The more I look at it, the more I think the same think about PublishTrimmed.

**Comment by vitek-karas** [comment]
> The more I look at it, the more I think the same think about PublishTrimmed.

I've seen several asks for FDD Trimming over time (TerraFx - if I remember correctly is one example, WinUI is another).

FDD Single-File is also pretty useful and I've seen it used. Definitely not as common as SCD though.

I personally don't buy any argument toward making R2R default to SCD - those two are completely orthogonal even functionally.

So the only two which I think we might consider are:
* NativeAOT - there I agree that it should default to SCD, I also don't think we will have FDD NativeAOT anytime soon, or even later
* Trimming - this one is interesting. I think we will eventually have FDD Trimmed, but it's not clear to me if we should sacrifice the by far more common use cases. What I don't like is that it would "fail by default" if we don't make it SCD.

**Comment by agocke** [comment]
>  less verbose --scd flag as a UX improvement

We already have `--sc`.

**Comment by nagilson** [comment]
@sbomer Thanks for bringing me in here.
> Please correct me if I'm wrong, but I believe that the behavior is the consequence of "use current rid for single-file" combined with "rid implies self-contained" 

This is correct. 

Net 6: (No Implicit RID):
`dotnet publish -p:PublishSingleFile=true` --> `SelfContained=false`

Net 7: (Implicit RID):
`dotnet publish -p:PublishSingleFile=true` --> `RuntimeIdentifier` --> `SelfContained=true`

For `PublishSingleFile` the default FDD behavior is OK, it won't fail. And that was the default behavior in NET 6.

**Comment by nagilson** [comment]
@richlander  
> I actually thought we implemented PublishSelfContained last version, but I guess not. 

It was implemented and made it into `7.0.200` -- works for me. The confusing part is the release page version is called `7.0.2` but it is `7.0.102`... which is not `7.0.200`... The SDK installation page is based on the runtime version, and does not communicate the SDK version, and we have given them feedback in the past that it should be more clear. Maybe my comment here:

> Close this issue [...] As well as one for PublishSelfContained.

Added to the confusion, I meant to say an issue for inference of these properties.

**Comment by richlander** [comment]
> > Also, we're never going to build Native AOT FDD
> 
> 
> 
> The more I look at it, the more I think the same think about PublishTrimmed. 

The two examples are not remotely similar.

**Comment by agocke** [comment]
They are. The only thing that matters is our likelihood of building them. I don't think the likelihood of building FDD trimming is any higher than AOT. I have even *more* use cases for AOT FDD than I do for trimming, e.g. plugin scenarios.

**Comment by richlander** [comment]
Interesting. I had never thought of plug-ins. That said, I still think Native AOT and FDD is a bad idea due to versioning issues and giving up on perf due to a boundary.

I know what the rules are for CoreCLR with plugins in terms of multiple runtime versions. Does Native AOT have the same rules or different? The native plug-in scenario only makes sense to me if the rules are more relaxed.

Contrast that with trimming. It is easy to see how we could provide benefit for FDD apps, particularly the ones that take a dependency on the big cloud SDKs.

**Comment by agocke** [comment]
> Contrast that with trimming. It is easy to see how we could provide benefit for FDD apps, particularly the ones that take a dependency on the big cloud SDKs.

I covered some of the basic problems with it here: https://github.com/dotnet/linker/issues/2609. I don't think it will ever be feasible. We likely couldn't ever statically guarantee safety because we wouldn't know the code paths that could be executed by the dynamically-loaded framework which could end up doing reflection against the target library.

**Comment by vitek-karas** [comment]
I don't want to use this issue to discuss the feasibility of FDD trimming. That said I don't think the problem of trimming one assembly only is the same as FDD trimming - with FDD trimming we would make an assumption that framework is fully annotated. With that you can do some things - not everything works, but the basic removal of obviously dead code should work just fine - and in a safe manner. (Servicing of the framework is a tricky question though... but that's too much detail right now).

**Comment by richlander** [comment]
We're going around in circles a bit, in large part because there is no obvious "right answer".

Let's make this a little easier:

- Sven: Make everything FDD by default, even if that means error. The win is a consistent product.
- Rich: Same as Sven except for scenarios that we're certain will be SCD only forever (which is Native AOT).
- Andy and Vitek: Choose FDD or SCD based on what's the most common choice today.

Is that fair?

**Comment by eerhardt** [comment]
I am _very_ against making PublishAot FDD by default. It would mean our new `dotnet new api -aot` template would need to put 2 properties in the csproj instead of 1. It just adds clutter/confusion in my mind. 

Properties inferring other properties makes total sense. PublishAot implies PublishTrimmed and PublishSingleFile. I could imagine a world where I want PublishAot but don’t trim anything. Is it surprising that adding PublishAot also flips PublishTrimmed? `dotnet publish` doesn’t trim (just like it isn’t self contained). Why is it OK for PublishTrimmed to get its non-default value when I PublishAot?

The same argument applies to SelfContained. If it isn’t explicitly specified, it can get defaulted to what makes the most sense.

**Comment by richlander** [comment]
I agree with you, however your "very against" is a bit overplayed in terms of how many properties we have in a project file. I don't think that's an important metric for this decision. My continued claim is that this conversation is primarily a CLI one (for "neutral" projects). We've added more properties over the last couple releases and I'm sure we'll add more.

**Comment by eerhardt** [comment]
> My continued claim is that this conversation is primarily a CLI one (for "neutral" projects).

1. When it comes to PublishAot, PublishTrimmed, and PublishSingleFile, we recommend devs put these properties in their csproj. For example, see https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-7-0#enable-trimming

2. I see no reason why the property defaulting behavior would be any different between the command line and the csproj. Passing `-p:PublishXXX` on the command line would default the same properties as setting `<PublishXXX>` in the csproj.

**Comment by richlander** [comment]
I meant something different. Clearly, the AOT template should be configured correctly to provide the intended results. I view near 100% of the discussion on this thread to center around trimming. That clearly wasn't about the AOT template, but about project files that don't have any of these properties in them and what just specifying `-p:PublishTrimming=true` via the CLR should do. I don't think anyone raised what the property on its own in a project file should do.

**Comment by eerhardt** [comment]
> what just specifying -p:PublishTrimming=true via the CLR should do. I don't think anyone raised what the property on its own in a project file should do.

That’s probably my mistake. My presupposition was that setting a property via the command line and setting a property in the .csproj results in the same behavior. I could have just as easily written the original issue with setting `<PublishedTrimmed>` in the csproj.

**Comment by richlander** [comment]
>  My presupposition was that setting a property via the command line and setting a property in the .csproj results in the same behavior.

They do.

> I could have just as easily written the original issue with setting <PublishedTrimmed> in the csproj.

But that wouldn't be nearly as compelling. Writing settings into a project file is a one and done activity. Typing things via the CLI is every time.

**Comment by jkotas** [comment]
> When it comes to PublishAot, PublishTrimmed, and PublishSingleFile, we recommend devs put these properties in their csproj.

These properties enable analyzers that warn about the form factor incompatibilities in the inner dev loop.

Setting these properties via command line is not a pit of success.

**Comment by agocke** [comment]
You have my position mostly right Rich, but the big reason I think we should make trimmed imply scd is that i strongly believe we'll never do FDD for either aot or trimming, and I find it obnoxious when dev tools have one possible interpretation of your command that will work, and then produce an error instead of doing it. 

The risk here is that I'm wrong and we eventually add FDD. But that risk is tempered by the likelihood that even if we add those configurations, they'll probably be a lot less popular.

**Comment by agocke** [comment]
That said, I agree that adding things to the project file once isn't a huge deal. Tbh I hate "Self Contained" build a lot more than some extra typing. Self contained build is *never* what I want and it just clutters the output and makes my artifacts directory ~100 MB larger than it should be.

**Comment by nagilson** [comment]
> My presupposition was that setting a property via the command line and setting a property in the .csproj results in the same behavior.

> > They do.

This is not always true. For example: `-p:Configuration=Release` and `<Configuration>Release</Configuration>`. One works, the later one doesn't.

**Comment by richlander** [comment]
I see significant diversity of opinion for FDD trimming on this thread on feasibility, value, and likelihood.

I'd like to offer the options that Sven and I offered. Can we get behind those and choose one? FWIW, I'm happy with either so not specifically trying to push the one I suggested. It's all just tradeoffs.

**Comment by eerhardt** [comment]
@richlander - can you explain why you are cutting the "Andy and Vitek" option from consideration?

> I see significant diversity of opinion for FDD trimming on this thread on feasibility, value, and likelihood.

My opinion is that even if FDD trimming was feasible, valuable, and likely, I would still want PublishAot, PublishTrimmed, and PublishSingleFile to be SCD by default.

**Comment by richlander** [comment]
I was acting on Jan's feedback, which is that we shouldn't be optimizing for the CLI but for project files for setting these properties. I agree with ath. If we're talking about project files, then a consistent model trumps reducing properties.

**Comment by agocke** [comment]
I'm fine with going forward with FDD-by-default.

**Comment by richlander** [comment]
BTW: A lot of this is my fault. I pushed implying SCD for some of the `Publish*` settings in .NET 7. It was clearly not well thought through.

**Comment by richlander** [comment]
I propose we make that change in Preview 2 and then course correct as needed. Fair?

That means all the `Publish*` settings will be FDD by default, except `PublishSelfContained`.

That leaves what to do with `SelfContained`. I think we should just undocument and call it good. I'm not sure that breaking it is useful. Note that this is separate from the CLI args, which should stay as-is.

**Comment by agocke** [comment]
> I think we should just undocument and call it good

We'll need a new command line argument then, and we'll need to change error messages

**Comment by agocke** [comment]
That is, you'll get an error message from AOT and trimming saying you need to enable self-contained. We need to instead say, PublishSelfContained or `-p:PublishSelfContained`

**Comment by richlander** [comment]
I think we're getting at CLI and project file UX being different.

The following seems fine to me:

```bash
dotnet publish -p:PublishTrimmed --sc
```

This seems like way too much:

```bash
dotnet publish -p:PublishTrimmed -p:PublishSelfContained
```

If we want to leave `SelfContained` alone, that's fine. I was trying to re-raise a suggestion you had raised earlier. Do you have another suggestion on how to approach that?

**Comment by agocke** [comment]
My preference is that we would change SelfContained to mean PublishSelfContained. If you want a self contained build, you'll need to execute a publish. Basically the same behavior, just behind a different verb.

**Comment by richlander** [comment]
I see. This change would mean that we are _for sure_ saying that the way to produce final prod apps with .NET is via `publish`. I think we're mostly there anyway. If no one else wants to raise a concern, I'm good with that. In fact, it might simplify the whole model.

@baronfel @dsplaisted

**Comment by marcpopMSFT** [comment]
It sounds like we're settling on a plan that is FDD by default always. That means introducing errors when aot, trimming, and singlefile correct?  Would `dotnet build --sc` also produce an error?

+1 to Andy's suggestion of SelfContained meaning PublishSelfContained

Would we need to coordinate changes to the api -aot template to add the PublishSelfContained property? Are there any other templates that would need to be changed (do we have any way of searching the templates for any of the three Publish properties that are affected)?

**Comment by richlander** [comment]
Mostly "yes" and great questions.

> That means introducing errors when aot, trimming, and singlefile correct?

- For .NET 6 apps, no.
- For .NET 7 apps, yes depending on how they are configured.
- For .NET 8 apps, no, since we'll change the templates.

> Would `dotnet build --sc` also produce an error?

With this proposal, yes. We'd move that to `dotnet publish` only.

> Would we need to coordinate changes to the api -aot template 

Yes.

This actually aligns quite closely with the `-r` change now that I think about it. It is just one step further.

**Comment by DamianEdwards** [comment]
I've read all the reasoning here, and while I don't fault the logic, I find the fact that `PublishAot` will now simply error without `PublishSelfContained` also being specified a strange UX for us to default to, especially when we seem to believe it's extremely unlikely we'd ever support AOT with FDD apps.

**Comment by richlander** [comment]
You are not wrong. That's why there were both "Sven" and "Rich" options.

However, following Jan's logic, this should not be an actual problem. Fixup your project file and it will work.

While my proposal is more pragmatic, I think Sven's proposal is perhaps stronger. We get to document that we offer exactly one model and the project files always look the same way. For example, SCD+Trimming look the same as SCD+AOT. That argument falls down a bit since some properties might infer trimming, but I'm choosing to ignore that.

**Comment by dsplaisted** [comment]
> My preference is that we would change SelfContained to mean PublishSelfContained. If you want a self contained build, you'll need to execute a publish. Basically the same behavior, just behind a different verb.

> I see. This change would mean that we are _for sure_ saying that the way to produce final prod apps with .NET is via `publish`. I think we're mostly there anyway. If no one else wants to raise a concern, I'm good with that. In fact, it might simplify the whole model.

This seems like a major change breaking change.  Maybe publish should be the way to get the final prod app, but I bet there are lots of people that don't do it that way today.

One thing that springs to mind is Exe to Exe project references.  It's somewhat of a niche scenario, but today you can have a self-contained app reference another self-contained app, and you can get a working copy of both in the output folder.  This wouldn't work with publish, since when you publish a project, you don't publish its project references.

**Comment by agocke** [comment]
> This wouldn't work with publish, since when you publish a project, you don't publish its project references.

It works fine though, you just need to pass `<ProjectReference Targets="...;Publish" />` instead of a plain ProjectReference. I'd much rather people do that since right now there's a large cliff where if you want just a self-contained build, that can work with no extra configuration, but if you want a single-file build you have to switch to Publish and all the associated configuration.

I'd rather we have one, well-documented way of doing this. And since publish is our gesture going forward for app deployment, we should group this all under publish.

**Comment by dsplaisted** [comment]
> It works fine though, you just need to pass `<ProjectReference Targets="...;Publish" />` instead of a plain ProjectReference. 

Are you sure that works?  There's a protocol for getting the files from a referenced project, and I don't think running the Publish target on the reference would mean that it would return the published files instead of the build output to the referencing project.

**Comment by richlander** [comment]
I don't have a strong viewpoint on this topic. Perhaps the two of you can hash out a proposal?

**Comment by agocke** [comment]
I don't have a ProjectReference sample handy so I can't test it, but I don't think it matters because I know this works:

```C#

<MSBuild Projects="$(RepoRoot)src/coreclr/tools/aot/crossgen2/crossgen2.csproj"
             Targets="Restore"
             Properties="MSBuildRestoreSessionId=$([System.Guid]::NewGuid())
              ;_IsPublishing=true" />

    <MSBuild Projects="$(RepoRoot)src/coreclr/tools/aot/crossgen2/crossgen2.csproj"
             Targets="Publish;PublishItemsOutputGroup"
             Properties="_IsPublishing=true">
      <Output TaskParameter="TargetOutputs"
              ItemName="_RawCrossgenPublishFiles" />
    </MSBuild>
```

And that's still a better alternative to using SelfContained ProjectReference. If we want to make sure Publish can work with ProjectReference I'd be fine with fixing it, but I'd also be fine with telling people to just use the MSBuild task. Again, one thing that works in all circumstances is better than multiple things that only work some of the time.

**Comment by OptiStrat** [comment]
> 



> > The more I look at it, the more I think the same think about PublishTrimmed.
> 
> I've seen several asks for FDD Trimming over time (TerraFx - if I remember correctly is one example, WinUI is another).
> 

@richlander @agocke @vitek-karas as part of this thread discusses the usefulness of FDD Trimming and the likelihood to see this feature implemented: I have a scenario where I would have to xcopy-deploy a relatively large set of relatively simple unpackaged WinUI apps. The C#/WinRT projections alone add ~20 MB to every single app. With CsWinRT 2.0, these projections are trimmable, and it would be great if FDD could benefit from this.

@agocke are the challenges described in [linker/issues/2609](https://github.com/dotnet/linker/issues/2609) applicable to my scenario as well, or do you think it would be easier to support?

**Comment by vitek-karas** [comment]
The issue https://github.com/dotnet/linker/issues/2609 was actually motivated by WinUI projections assembly (it just doesn't say so specifically). But it's ultimately somewhat different problem.

The expectation is that WinUI assemblies are simple in that they have only references to framework and not user code assemblies - so they act as "leafs" in the analysis in that sense (assuming FDD trimming like picture). So supporting trimming for these would probably be a bit easier, but I would expect it to hit most of the problems described in https://github.com/dotnet/linker/issues/2139 regardless.

The problems described in https://github.com/dotnet/linker/issues/2609 are specifically to support a hybrid world where only some assemblies are trimmed AND not everything is analyzed. The trimmer already supports the case where not everything is trimmed, but it does assume that it will analyze everything. https://github.com/dotnet/linker/issues/2609 tries to explore if it's possible to not analyze most of the app if only a small set of assemblies are to be trimmed. That so far proves to be rather difficult.

**Comment by eerhardt** [comment]
Retested with SDK v`8.0.200-preview.23620.12` and the repro steps in the original post now work correctly. Closing.
